flax/linen/module.py

```diff
@@ -705,6 +705,17 @@ def __init_subclass__(cls, **kwargs: Any) -> None:
     cls._parent_ref = None # type: ignore[attr-defined]
     cls.parent = ParentDescriptor() # type: ignore[attr-defined]
 
+    # Force all sub-classes to call `super(cls, self).__post_init__()`
+    def force_super_post_init(post_init_method):
+      @functools.wraps(post_init_method)
+      def new_post_init_method(self):
+        post_init_method(self)
+        if not hasattr(self, '_id') and hasattr(super(cls, self), '__post_init__'):
+          super(cls, self).__post_init__()
+      return new_post_init_method
+
+    cls.__post_init__ = force_super_post_init(cls.__post_init__) # type: ignore
```

Comment:
isn't this presuming that a cls.__post_init__ exists on this particular subclass?  I think the logic of your wrapper will do the right thing still (due to this recursive looking up the method resolution order)...

Reply:
I added an `if` condition to check if it has a `post_init`